store/src/java/com/zimbra/cs/event/SolrEventStore.java

```diff
@@ -229,6 +233,131 @@ private String getAccountFilter(String accountId) {
+
+        SolrQuery solrQuery = new SolrQuery();
+        solrQuery.setQuery(new MatchAllDocsQuery().toString());
+        solrQuery.setRows(0);
+        solrQuery.addFilterQuery(searchContactInSenderField.toString());
+        solrQuery.addFilterQuery(filterByEventTypes.build().toString());
+
+        if (solrHelper.needsAccountFilter()) {
+            solrQuery.addFilterQuery(getAccountFilter(accountId));
+        }
+
+        solrQuery.setFacet(true);
+        solrQuery.addFacetField(LuceneFields.L_EVENT_TYPE);
+
+        QueryResponse response = (QueryResponse) solrHelper.executeRequest(accountId, solrQuery);
+        if(response.getResults().getNumFound() <= 1) {
+            return 0.0;
+        }
+
+        Map<String, Long> eventFacetResults = Maps.newHashMap();
+        response.getFacetFields().get(0).getValues().forEach(t -> eventFacetResults.put(t.getName(), t.getCount()));
+        if(eventFacetResults.containsKey(Event.EventType.RECEIVED.name()) && eventFacetResults.containsKey(Event.EventType.READ.name())) {
+            Double numberOfReceivedEmails = Double.valueOf(eventFacetResults.get(Event.EventType.RECEIVED.name()));
+            Double numberOfReadEmails = Double.valueOf(eventFacetResults.get(Event.EventType.READ.name()));
+            if(numberOfReadEmails.isNaN() || numberOfReadEmails == 0
+                    || numberOfReceivedEmails.isNaN() || numberOfReceivedEmails == 0) {
+                return 0.0;
+            }
+            return (numberOfReadEmails / numberOfReceivedEmails);
+        }
+        return 0.0;
+    }
+
+    @Override
+    public Double getAvgTimeToOpenEmail(String contact) throws ServiceException {
+        try {
+            SelectStream seenEventSelectStream = getSelectStreamForEventTypeWithSearchStream(Event.EventType.SEEN, contact);
+            SelectStream readEventSelectStream = getSelectStreamForEventTypeWithSearchStream(Event.EventType.READ, contact);
+            FieldEqualitor messageIdEqualitor = new FieldEqualitor(LuceneFields.L_EVENT_MESSAGE_ID);
+            FieldEqualitor senderEqualitor = new FieldEqualitor(SolrEventDocument.getSolrQueryField(Event.EventContextField.SENDER));
+            MultipleFieldEqualitor equalitor = new MultipleFieldEqualitor(messageIdEqualitor, senderEqualitor);
+            Double delta = 0.0;
+            int count = 0;
+            try (JoinStream joinStream = new InnerJoinStream(seenEventSelectStream, readEventSelectStream, equalitor)) {
+                joinStream.setStreamContext(getStreamContext());
+                joinStream.open();
+                Tuple tuple = joinStream.read();
+                while (!tuple.EOF) {
+                    Date seenDate = tuple.getDate("seen_timestamp");
+                    Date readDate = tuple.getDate("read_timestamp");
+                    delta = delta + (readDate.getTime() - seenDate.getTime());
```

Comment:
I think we should ignore cases when the delta is below a certain threshold. This is because it's currently possible for a SEEN and READ event to be triggered back-to-back in some scenarios. Maybe if delta is below, say, 500ms, skip it?

Reply:
I think with the incremental model we should not ignore it. I think if we want to ignore it then we can handle that in ContactAnalytics class when it gets the value back from the model.